Implement mass machine type transition job WIP

prnaraya opened the pull request:
The mass machine type job is implemented with a Docker image built locally from a golang file, and creating and applying a Job using that image. Working on adding way for user to specify namespace(s) to run the Job on rather than running on all namespaces, and a flag for the Job to force restart all running vms when changing the machine type instead of letting the user do it manually.

prnaraya committed:
WIP commit, adding the base functions for the job and the informers

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Reorganized the functions/files and added a completion count function.
After going back and forth on whether to implement using nested jobs or not, I settled for
having a single job, with the calculation of the completion count being done in a go function.
There are also some sections still commented out as they are WIP, notably dealing with the actual
machine type strings themselves.

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Updated README

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Implemented job using docker image created from golang code

Signed-off-by: prnaraya <prnaraya@redhat.com>

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+		return err
+	}
+	for _, vm := range vmList.Items {
+		machineType := vm.Spec.Template.Spec.Domain.Machine.Type
+		machineTypeSubstrings := strings.Split(machineType, "-")
+		version := machineTypeSubstrings[2]
```
Are we sure that every VMI would have a machine type of this format?
I've created a `/examples/vmi-fedora.yaml` vmi and saw:
```yaml
spec:
  domain:
    machine:
      type: q35
```

In this case this job would crash.
I'd recommend to validate it's the right format.

iholder101 replied:
I do see 

```yaml
  machine:
    type: pc-q35-rhel9.2.0
```

under status though

iholder101 replied:
Okay so after some investigation I think I understand what needs to be done exactly.
Let's take the following as an example:
```yaml
spec:
  domain:
    machine:
      type: q35
status:
  machine:
    type: pc-q35-rhel8.2.0
```

So, what would need to be done is:
1. Parse `status.machine.type`, specifically the last bit (split by `-` like you did, take the 3rd string, split by `.`, take the first string (which in this case is `8`), then parse it to a number and check if it's `<= 8`, like in this case)*.
2. If the machine type is of rhel8 or below, update `spec.domain.machine.type` to a RHEL9 version**.
3. Add the label.

Open questions:
\* Can we count on the fact that all machine types at the status level look like `pc-q35-rhel8.2.0`? Do we support other machine types with different formats?
\** Which RHEL version should we update to? `9.0.0`? `9.2.0`?

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+		machineTypeSubstrings := strings.Split(machineType, "-")
+		version := machineTypeSubstrings[2]
+		
+		if strings.Contains(version, "rhel") && version < "rhel9.0.0" {
```
What did you try to achieve with `version < "rhel9.0.0"`? Did you mean `version != "rhel9.0.0"`?
What if the version is `"rhel9.2.0"` or something similar?
Do all machine types have a version?

prnaraya replied:
Basically the requirements for updating the machine types is updating the machine type of any VM with a machine type that is prior to RHEL 9. There are some other formats of machine type I believe so I will need validate to prevent this from crashing though.

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+
+func addWarningLabel (virtCli kubecli.KubevirtClient, vm *k6tv1.VirtualMachine) {
+	addLabel := fmt.Sprint(`{"metadata": {"labels": {"restart-vm-required": "true"}}}}}`)
+	virtCli.VirtualMachine(vm.Namespace).Patch(vm.Name, types.StrategicMergePatchType, []byte(addLabel), &k8sv1.PatchOptions{})
```
WDYT about catching an error and logging it if exists?

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+	
+	removeLabel := fmt.Sprint(`[{"op": "remove", "path": "metadata/labels/restart-vm-required"}]`)
+	virtCli.VirtualMachine(vmi.Namespace).Patch(vmi.Name, types.StrategicMergePatchType, []byte(removeLabel), &k8sv1.PatchOptions{})
+	needsRestart--
```
This occurs for every VMI that's being deleted.
So it can decrease the counter for a VMI that didn't need any updates in the first place.

iholder101 commented on main.go:
```diff
@@ -3,14 +3,27 @@ package mass_machine_type_transition
+	// on any VMs before exiting the job
+	
+	for (needsRestart > 0) {
+	}
```
This would busy-wait for the variable to become zero, which is pretty costly in terms of performance. A better approach would be to sleep for a few seconds between iterations. Or even better, use a channel to indicate when to wake up and exit.

But, currently, this mechanism is broken since the counter decreases when every VMI is deleted, not necessarily a relevant one. I think you need to build a structure of VMIs that should be restarted (in `updateMachineTypes`) to really track the relevant ones. Then, when the structure is empty, you can close a channel that would indicate to exit the job.

prnaraya replied:
Yeah, this was a temporary fix that I implemented just so I had something there to use when testing against a small cluster, but I wasn't exactly sure how to implement this in a better way.

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+	needsRestart++
+}
+
+func removeWarningLabel(obj interface{}) {
```
no need to remove a label from a deleted VMI.
I think the main purpose of this event handler is to update that we have one less VMI to track

iholder101 commented on informers.go:
```diff
@@ -23,14 +29,64 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
 	listWatcher := cache.NewListWatchFromClient(virtCli.RestClient(), "virtualmachineinstances", k8sv1.NamespaceAll, fields.Everything())
 	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+	
+	vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs {
+		DeleteFunc: removeWarningLabel,
+	})
```
WDYT about adding an event handler for newly created VMIs?
That way if a new VMI is being created with an unsupported machine type it could be added to the list of VMIs that need update, and a label could be added.

Although I'm not entirely sure it's necessary.

prnaraya replied:
I could add that but I am also not sure if it is something that we would need.

iholder101 replied:
Let's skip that for now. we can always add that in follow-ups

iholder101 reviewed (commented):
Thanks @prnaraya! This is surely a step forward!

Left some comments below.

prnaraya committed:
Adding vmiList to track VMs that need to be restarted
and a stop channel for the job to wait until vmiList
is empty

Signed-off-by: prnaraya <prnaraya@redhat.com>

iholder101 commented on informers.go:
```diff
@@ -49,9 +54,12 @@ func updateMachineTypes(virtCli kubecli.KubevirtClient) error {
+			return nil
+		}
 		
 		if strings.Contains(version, "rhel") && version < "rhel9.0.0" {
```
We need to change `version < "rhel9.0.0"`.
Basically, we need to check if the x-stream version (9 in this case) is >8. In other words, if it's less or equal to 8 (e.g. 8.4.0) it needs to be updated

prnaraya replied:
I am not sure I understand. Since we only want to update machine types with versions prior to the specified version (in this case RHEL 9), wouldn't a lexographical comparison of the versions be sufficient? Unless there is a format of machine type that I am missing.

In this case, "rhel8.4.0" < "rhel9.0.0" returns true, which satisfies the conditions I believe.

iholder101 replied:
Yup, you have a point.

As long as we're sure this is indeed always the format, your approach is completely valid. Saying that, it is never too bad to add some more validations, as this can go silently wrong if something changes (what if the string is empty for some reason? or if the format changes slightly, like `rhel-8.0.0` instead of `rhel8.0.0`).

But again, if you're certain that this is always the format, feel free to leave it as it :+1:

iholder101 commented on informers.go:
```diff
@@ -62,17 +70,25 @@ func updateMachineTypes(virtCli kubecli.KubevirtClient) error {
-	virtCli.VirtualMachine(vm.Namespace).Patch(vm.Name, types.StrategicMergePatchType, []byte(addLabel), &k8sv1.PatchOptions{})
-	needsRestart++
+	virtCli.VirtualMachineInstance(vm.Namespace).Patch(context.Background(), vm.Name, types.StrategicMergePatchType, []byte(addLabel), &k8sv1.PatchOptions{})
+	if err != nil {
```
Where is the `err` variable declared / initialized?
I think you need `err := virtCli.VirtualMachineInstance(vm.Namespace).Patch(....)`

prnaraya committed:
Adding restartnow flag and implemented logic to restart all affected running vmis
immediately. Also added the base for the namespace and potentially labels flags
to implement filtering of VMs in the future.

Signed-off-by: prnaraya <prnaraya@redhat.com>

iholder101 commented on main.go:
```diff
@@ -3,14 +3,24 @@ package mass_machine_type_transition
 	}
-
+	
+	stopCh := make(chan struct{})
+	go vmiInformer.Run(stopCh)
```
You can use `exitJob` for that purpose too

```suggestion
	exitJob = make(chan struct{})
	go vmiInformer.Run(exitJob)
```

iholder101 replied:
This will also initialize `exitJob` which is never initialized currently

iholder101 commented on main.go:
```diff
@@ -3,14 +3,24 @@ package mass_machine_type_transition
+	go vmiInformer.Run(stopCh)
+	
+	
+	updateMachineTypes(virtCli)
```
Please handle the error (by either printing it to stdout or logging it)

iholder101 commented on informers.go:
```diff
@@ -23,14 +34,89 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+		return err
+	}
+	for _, vm := range vmList.Items {
+		machineType := vm.Spec.Template.Spec.Domain.Machine.Type
```
This needs to be fetched from status instead of spec

prnaraya replied:
I've tried to fetch this but neither [VirtualMachineStatus](https://pkg.go.dev/kubevirt.io/api/core/v1#VirtualMachineStatus) nor [VirtualMachineInstanceStatus](https://pkg.go.dev/kubevirt.io/api/core/v1#VirtualMachineInstanceStatus) have a field for machine types.

iholder101 replied:
You can find it [here](https://github.com/kubevirt/kubevirt/blob/v1.0.0-rc.0/staging/src/kubevirt.io/api/core/v1/schema.go#L387). As mentioned in [this comment](https://github.com/prnaraya/mass-machine-type-transition/pull/2#discussion_r1212794849), this is how it looks like:
```yaml
spec:
  domain:
    machine:
      type: q35
status:
  machine:
    type: pc-q35-rhel8.2.0
```

prnaraya replied:
I looked into this a little more, and I believe the reason why the VMI you created has just "q35" in the machine type in spec is because that is the alias for the default machine type (which is the most recent RHEL version I believe) if a machine type is not explicitly specified in the yaml. I created a VM where I explicitly set the machine type, and was able to see the full machine type string in the Spec for both the VM and the VMI. This also works if you modify examples/vmi-fedora.yaml to have an explicitly stated machine type.

That is to say, I think it is still correct to retrieve the machine type from the VM spec, and there already exists logic that will catch any machine types that do not fit the "pc-q35-rhelx.x.x" version, so VMs with a machine type of just "q35" will be ignored.

iholder101 replied:
> so VMs with a machine type of just "q35" will be ignored

In my opinion this is a bug.
What if the VMIs I've created were created at a time where `8.2.0` was the default and latest version? Then you're going to ignore these, although they need to be updated.

Am I missing anything?

prnaraya replied:
Ah I see what you mean. I still am not sure how to access the VMI status.machine.type because I just don't see it in any of the documentation anywhere, but it might be good then in the case of the machine type being `q35` to update the machine type anyway to be safe.

iholder101 commented on informers.go:
```diff
@@ -1,14 +1,25 @@
 
+// using this as a const allows us to easily modify the program to update if a newer version is released
+// we generally want to be updating the machine types to the most recent version
+const latestVersion = "rhel9.2.0"
```
nit:

```suggestion
const latestMachineTypeVersion = "rhel9.2.0"
```

iholder101 commented on informers.go:
```diff
@@ -23,14 +34,89 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+	if err != nil {
+		return err
+	}
+	vmiList = append(vmiList, vm.Name)
```
Please all add the VM's namespace, that way it's a unique identifier.
BTW, you have `MetaNamespaceKeyFunc` function from Kubernetes (that you can import) that returns that string for you (i.e. `<namespace>/<name>`)

iholder101 commented on informers.go:
```diff
@@ -23,14 +34,89 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
 	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+	
+	vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs {
+		DeleteFunc: removeWarningLabel,
```
I'm not entirely sure, but I think that reboots might not delete the vmi entirely, but rather leave in a `Succeeded` state. Just to be on the safe side, I'd also add an update handler and re-use the same function you already have.

I would also add a small validation to `removeWarningLabel()` to make sure the VMI is either deleted or in a non-running phase.

```suggestion
		DeleteFunc: removeWarningLabel,
		UpdateFunc: func(_, newObj interface{}) {
			removeWarningLabel(newObj)
		},
```

iholder101 commented on informers.go:
```diff
@@ -23,14 +34,89 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+		return
+	}
+	
+	//check if deleted VMI is in list of VMIs that need to be restarted
+	vmiIndex := searchVMIList(vmi.Name)
+	if  vmiIndex == -1 {
+		return
+	}
+	
+	// remove deleted VMI from list
+	vmiList = append(vmiList[:vmiIndex], vmiList[vmiIndex+1:]...)
```
Not that it's critical, but I think this could have been much easier with a map.

The map could be of type `vmisPendingUpdate := make(map[string]struct{})`.
Then, to search: `_, exists := vmisPendingUpdate[<vm>]`
To delete: `delete vmisPendingUpdate[vmisPendingUpdate, <vm>`

iholder101 reviewed (commented):
Awesome progress @prnaraya! Well done!

prnaraya committed:
Addressing Itamar's comments

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Adding env variables for restart now flag and namespace flag

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Merge branch 'restart-now-flag' into dev

Signed-off-by: prnaraya <prnaraya@redhat.com>

iholder101 commented on informers.go:
```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+			}
+			
+			// add label to running VMs that a restart is required for change to take place
+			if vm.Status.Ready {
```
```suggestion
			if vm.Status.Created {
```

There's a slight difference between `Created`, which means that a VMI object exists for this VM, and `Ready`, which means that the VMI is up and ready (e.g. booted, connected, etc).

If we use `Ready` here we might miss a VM that had just started but did not fully boot yet. However, at this point, the machine type is already determined.

iholder101 commented on informers.go:
```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+func handleVmiDeletion(obj interface{}) {
+	vmi, ok := obj.(*k6tv1.VirtualMachineInstance)
+	if !ok {
+		return
```
Please log / print an error in this case, since it never should happen

iholder101 commented on informers.go:
```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+	
+	vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs {
+		UpdateFunc: handleVmiUpdate,
```
Sorry but it seems I've mislead you :(
I've tested this, and it seems that whenever a VMI is either stopped or restarted the VMI object is being deleted:
```
  metadata:
    deletionTimestamp: "2023-06-14T06:13:01Z"
```

So I guess we can drop the update handler after all

iholder101 commented on informers.go:
```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
-	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+func addWarningLabel (virtCli kubecli.KubevirtClient, vm *k6tv1.VirtualMachine) error {
+	addLabel := fmt.Sprint(`{"metadata": {"labels": {"restart-vm-required": "true"}}}}}`)
+	_, err := virtCli.VirtualMachineInstance(vm.Namespace).Patch(vm.Name, types.StrategicMergePatchType, []byte(addLabel), &k8sv1.PatchOptions{})
```
IMO we better add the label to the VM object instead of the VMI.
There are two reasons for that:
1. Technically, a VMI object cannot be restarted, only a VM object
2. The user-facing object is the VM object. This is meaningful, since most users are aware (in terms of monitoring, querying data, etc) of the VM object, and the VMI object is treated as an "implementation detail", similarly to the virt-launcher pod.

WDYT?

prnaraya replied:
This makes sense to me. I think this might have been what I was originally planning to do (I can't remember why I decided not to but I can't think of a good reason for it).

iholder101 commented on informers.go:
```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
+	removeWarningLabel(vmi)
+}
+
+func removeWarningLabel(vmi *k6tv1.VirtualMachineInstance) {
```
nit: I'd make the function name a bit more accurate, since it doesn't only remove the label, but also removed the VMI from the map, ends the job if necessary, etc.

WDYT about something like `handleDeletedVmi`?

iholder101 commented on main.go:
```diff
@@ -1,16 +1,49 @@
-	vmiInformer, err := getVmiInformer()
+	var err error
+	// update restartNow if env is set
+	restartEnv, exists := os.LookupEnv("RESTART_NOW")
+	if exists {
+		restartNow, err = strconv.ParseBool(restartEnv)
+		if err != nil {
+			fmt.Println(err)
+		}
+	}
+	
+	// update namespace if env is set
+	namespaceEnv, exists := os.LookupEnv("NAMESPACE")
+	if exists {
+		namespace = namespaceEnv
+	}
+	
```
Awesome! :+1:

iholder101 commented on main.go:
```diff
@@ -1,16 +1,49 @@
 	}
-
+	
+	exitJob = make(chan struct{})
```
This is already initialized above:

```go
var (
	vmisPendingUpdate = make(map[string]struct{})
	exitJob = make(chan struct{})
...
```

iholder101 commented on mass_machine_type_transition_job.yaml:
```diff
@@ -0,0 +1,14 @@
+kind: Job
+metadata:
+  name: mass-machine-type-transition
+spec:
+  template:
+    metadata:
+      name: mass-machine-type-transition
+    spec:
+      containers:
+      - name: machine-type-transition
+        image: machine-type-transition:latest
+        imagePullPolicy: IfNotPresent
+      restartPolicy: OnFailure
```
Please add the environment variables with default values to the job manifest. It will help guide users to use them properly.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: mass-machine-type-transition
spec:
  template:
    metadata:
      name: mass-machine-type-transition
    spec:
      containers:
      - name: machine-type-transition
        image: machine-type-transition:latest
        imagePullPolicy: IfNotPresent
        env:
        - name: NAMESPACE
          value: ""  # default is having no selector
        - name: RESTART_NOW
          value: false
      restartPolicy: OnFailure
```

iholder101 commented on main.go:
```diff
@@ -1,16 +1,49 @@
-	vmiInformer, err := getVmiInformer()
+	var err error
+	// update restartNow if env is set
+	restartEnv, exists := os.LookupEnv("RESTART_NOW")
```
WDYT?

```suggestion
	restartEnv, exists := os.LookupEnv("FORCE_RESTART")
```

iholder101 reviewed (commented):
Awesome progress @prnaraya!! Well done! :clap:

prnaraya committed:
Removed update function for the VMIs as it is not necessary
Added environment variables to the job yaml

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Restructured files to move non-informer functions into their own file.

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Updating machine type of all VMs who have machine type of "q35"

Signed-off-by: prnaraya <prnaraya@redhat.com>

prnaraya committed:
Update the addWarningLabel to apply the label to the VM rather than the VMI
Also updated HandleDeletedVmi to remove the label from the VM accordingly.

Signed-off-by: prnaraya <prnaraya@redhat.com>
